File: csharp/Betfair.ESAClient/Betfair.ESAClient/Cache/LevelPriceSizeLadder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Betfair.ESAClient.Cache
{
    /// <summary>
    /// A level price size ladder with copy on write snapshot
    /// </summary>
    public class LevelPriceSizeLadder
    {
        /// <summary>
        /// Dictionary of level to LevelPriceSize
        /// </summary>
        private readonly SortedDictionary<int, LevelPriceSize> _levelToPriceSize = new SortedDictionary<int, LevelPriceSize>();
        private IList<LevelPriceSize> _snap = LevelPriceSize.EmptyList;

        public IList<LevelPriceSize> OnPriceChange(bool isImage, List<List<double?>> prices)
        {
            if (isImage)
            {
                //image is replace
                _levelToPriceSize.Clear();
            }

            if (prices != null)
            {
                //changes to apply
                foreach (List<double?> price in prices)
                {
                    LevelPriceSize levelPriceSize = new LevelPriceSize(price);
                    //keep zero's in the ladder as it is fixed depth
                    _levelToPriceSize[levelPriceSize.Level] = levelPriceSize;
                }
            }

            if (isImage || prices != null)
            {
                //update snap on image or if we had cell changes
                _snap = new List<LevelPriceSize>(_levelToPriceSize.Values);
            }
            return _snap;
        }
    }
}

File: java/client/src/main/java/com/betfair/esa/client/cache/util/LevelPriceSizeLadder.java
package com.betfair.esa.client.cache.util;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/** A level price size ladder with copy on write snapshot */
public class LevelPriceSizeLadder {
    private final Map<Integer, LevelPriceSize> levelToPriceSize = new TreeMap<>();
    private List<LevelPriceSize> snap = Collections.emptyList();

    public List<LevelPriceSize> onPriceChange(boolean isImage, List<List<Double>> prices) {
        if (isImage) {
            // image is replace
            levelToPriceSize.clear();
        }

        if (prices != null) {
            // changes to apply
            for (List<Double> price : prices) {
                LevelPriceSize levelPriceSize = new LevelPriceSize(price);
                levelToPriceSize.put(levelPriceSize.getLevel(), levelPriceSize);
            }
        }

        if (isImage || prices != null) {
            // update snap on image or if we had cell changes
            snap = new ArrayList<>(levelToPriceSize.values());
        }
        return snap;
    }
}
